### PR TITLE
Pull Request 3: Persisted Favoriting using Realm

### PR DESCRIPTION
**Pull request by rufusnguyen**
Hi @chug2k, 

Please help me to review Pull Request 3. 

For Pull Request 4: After adding pod install for Firebase, my workspace can't be built due to lack of something and then I totally forgot it. I'm still working on it. I will commit it tomorrow! ;(

### CODE REVIEW

File: Select_Flick_Week2/FavoriteMovieDTO.swift [hunk 1]
```diff
@@ -9,7 +9,11 @@
 import Foundation
 import RealmSwift
 
-class FavoriteMovieDTO : Object {
-    dynamic var username:String!
-    dynamic var favouriteMovies:[Int:Bool]!
+class FavoriteMovieDTO: Object {
```
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieDTO.swift:12]: DTO là gì?

File: Select_Flick_Week2/FavoriteMovieDTO.swift [hunk 2]
```diff
@@ -9,7 +9,11 @@
...
+    dynamic var movieId:Int = 0
+    dynamic var isLike:Bool = true
```
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieDTO.swift:14]: initializing to false seems safer.

File: Select_Flick_Week2/FavoriteMovieService.swift [hunk 3]
```diff
@@ -10,23 +10,59 @@ import Foundation
...
 class FavoriteMovieService {
+
+    static func getFavoriteMovies(username: String) -> UserFavoriteMoviesDTO {
+        let realm = try! Realm()
+        let predicate = NSPredicate(format: "username = %@", username)
+        let favoriteMovies = realm.objects(UserFavoriteMoviesDTO.self).filter(predicate)
+        if favoriteMovies.count == 0 {
+            return UserFavoriteMoviesDTO()
+        } else {
+            return (favoriteMovies.first)!
```
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieService.swift:21]: small note: don't need `( )`

Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieService.swift:21]: i'm confused what this method returns, actually. why does it sometimes return the DTO and sometimes a movie?

File: Select_Flick_Week2/FavoriteMovieService.swift [hunk 4]
```diff
@@ -10,23 +10,59 @@ import Foundation
...
         if favoriteMovies.count == 0 {
             // add new
+            var userFavMovies = UserFavoriteMoviesDTO()
+            userFavMovies.username = username
+            var movie = FavoriteMovieDTO()
+            movie.movieId = movieId
+            movie.isLike = true
+            movie.transactionDate = NSDate()
+            userFavMovies.favouriteMovies.append(movie)
+            //try! realm.commitWrite() {
```
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieService.swift:39]: remove comment :)

File: Select_Flick_Week2/FavoriteMovieService.swift [hunk 5]
```diff
@@ -10,23 +10,59 @@ import Foundation
...
+            try! realm.commitWrite()
+            
         } else {
-            var favoriteMovies = favoriteMovies.first?.favouriteMovies
-            // update
-            if favoriteMovies!.indexForKey(movieId) != nil {
-                favoriteMovies![movieId] = !(favoriteMovies![movieId]!)
-            } else {
-
+            var isFound:Bool = false
```
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieService.swift:45]: `var isFound: Bool = false`

File: Select_Flick_Week2/FavoriteMovieService.swift [hunk 6]
```diff
@@ -10,23 +10,59 @@ import Foundation
...
+            for movie in (favoriteMovies.first?.favouriteMovies)! {
```
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieService.swift:46]: don't need `( )`

File: Select_Flick_Week2/FavoriteMovieService.swift [hunk 7]
```diff
@@ -10,23 +10,59 @@ import Foundation
...
-            
-            
+            if isFound == false {
+                let movie = FavoriteMovieDTO()
+                movie.movieId = movieId
+                movie.isLike = true
+                movie.transactionDate = NSDate()
+                favoriteMovies.first?.favouriteMovies.append(movie)
+            }
+            //try! realm.write() {
```
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieService.swift:61]: clean up :)

File: Select_Flick_Week2/FavoriteMovieService.swift [hunk 8]
```diff
@@ -10,23 +10,59 @@ import Foundation
...
+            var test = favoriteMovies.first?.favouriteMovies
```
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieService.swift:62]: tại sao `var test`?

File: Select_Flick_Week2/MovieCell.swift [hunk 9]
```diff
@@ -8,19 +8,33 @@
 
 import UIKit
 import MGSwipeTableCell
+protocol LikeMovieCellDelegate {
```
Comment (chug2k) [inline Select_Flick_Week2/MovieCell.swift:11]: 👍

File: Select_Flick_Week2/MovieCell.swift [hunk 10]
```diff
@@ -8,19 +8,33 @@
...
 class MovieCell: MGSwipeTableCell {
     @IBOutlet weak var poster: UIImageView!
     @IBOutlet weak var nameLabel: UILabel!
     @IBOutlet weak var yearLabel: UILabel!
     @IBOutlet weak var voteCount: UILabel!
     @IBOutlet weak var overview: UILabel!
     @IBOutlet weak var rating: UILabel!
+    var isLike:Bool = false
+    var movie:Movie!
+    var delegateClick: LikeMovieCellDelegate!
```
Comment (chug2k) [inline Select_Flick_Week2/MovieCell.swift:24]: this is probably better as a `?`

File: Select_Flick_Week2/MovieCell.swift [hunk 11]
```diff
@@ -8,19 +8,33 @@
...
     
     override func prepareForReuse() {
         poster.image = nil
+        likeBtn.setBackgroundImage(nil, forState:  .Normal)
```
Comment (chug2k) [inline Select_Flick_Week2/MovieCell.swift:28]: 👍

File: Select_Flick_Week2/MovieCell.swift [hunk 12]
```diff
@@ -8,19 +8,33 @@
...
     }
     @IBOutlet weak var likeBtn: UIButton!
     @IBAction func clickLike(sender: AnyObject) {
-        likeBtn.setBackgroundImage(UIImage(named: "heart_true"), forState: UIControlState.Normal)
+        if isLike == false {
```
Comment (chug2k) [inline Select_Flick_Week2/MovieCell.swift:32]: in style, it's usually easier to write this as 

```
if isLIke { 
...
} else {
...
}
```

**Comment by chug2k** [comment]
Good job implementing like _and_ unlike. :)

File: Select_Flick_Week2/FavoriteMovieService.swift (hunk 6 above)
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieService.swift:46]: I think this logic is a little hard to follow because `isFound` is a confusing terminology. Maybe something is `isFavorited`?

File: Select_Flick_Week2/FavoriteMovieService.swift [hunk 13]
```diff
@@ -10,23 +10,59 @@ import Foundation
...
+                let m = movie as! FavoriteMovieDTO
+                if m.movieId == movieId {
+                    m.isLike = !movie.isLike
+                    m.transactionDate = NSDate()
+                    isFound = true
+                }
             }
-            
-            
+            if isFound == false {
```
Comment (chug2k) [inline Select_Flick_Week2/FavoriteMovieService.swift:54]: `if !isFound`

File: Select_Flick_Week2/FavoriteMovieDTO.swift (hunk 1 above)
Comment (rufusnguyen) [inline Select_Flick_Week2/FavoriteMovieDTO.swift:12]: In .NET mindset, DTO object is an model which coordinate directly with Db. 
I consider Realm is our DBMS, FavoriteMovieDTO is exactly a schema of table. Let me investigate the similar term in Swift. 👯

File: Select_Flick_Week2/FavoriteMovieDTO.swift (hunk 2 above)
Comment (rufusnguyen) [inline Select_Flick_Week2/FavoriteMovieDTO.swift:14]: my bad behavior again!
